### DISCUSSION TITLE
How to show results to participants?

### DISCUSSION THREAD

**Discussion by ler93** [jspsych/jsPsych#2277] [Q&A] [answered]
Hi all. I would like to show to participants the percentage of correct answers at the end of their performance in a raven test. Is there any way to do it? Any reference link?

Thanks!!

**Comment by becky-gilbert** [comment, marked as answer]
Hi @ler93, have you seen the example of this in the "data aggregation" step of the [jsPsych simple RT task tutorial](https://www.jspsych.org/7.0/tutorials/rt-task/#part-13-data-aggregation)? The trick is to use a [dynamic parameter](https://www.jspsych.org/7.0/overview/dynamic-parameters/) to present the percentage correct text to the participant (since this is information that you don't know in advance and need to generate _during_ the experiment) and to use the [jsPsych data manipulation functions](https://www.jspsych.org/7.0/overview/data/#aggregating-and-manipulating-jspsych-data) inside your dynamic parameter function to get the trials you want to summarize and calculate the performance measures. 

I hope that helps. Let us know if you get stuck on this and need any further guidance.

**Reply by ler93** [reply]
Hi, @becky-gilbert! I don't know why it doesn't return a numerical data at the end of the experiment (says "You answered correctly in **NaN**% of responses..."). I copy part of the code below: 

 ```
var timeline = [];

      var raven35 = {
        type: 'image-button-response',
        choices: 'Raven35_complete.png',
        options: ['1', '2', '3', '4', '5', '6', '7', '8'],
        task: 'raven',
        correct_response: '1'
      };
      timeline.push (raven35);
      
      var show_results = {
        type: 'html-keyboard-response',
        stimulus: function () {
          var trials = jsPsych.data.get(). filter ({task: 'raven'});
          var correct_trials = trial.filter ({correct: true});
          var accuracy = Math.round (correct_trials.count() / trial.count() * 100);
          
          return `<p> You answered correctly in $ {accuracy}% of the trials. </p>
          <p> Press any key to end the experiment. Thanks! </p> `;
        }
      };
      timeline.push (show_results);
      
  jsPsych.init ({
  timeline: timeline,
  });
```
Maybe there is an error in the code but I can't figure out what it is.

**Reply by becky-gilbert** [reply]
Ah, it looks like you're missing a few steps here. Specifically, the way that you've added `task: raven` and `correct_response: '1'` to the `raven35` trial won't work, because those things need to go inside the trial's `data` parameter. Also, you'll need to score each raven response as correct or incorrect when the trial is finished, which you can do in the trial's `on_finish` function. Have a look at [part 11](https://www.jspsych.org/7.0/tutorials/rt-task/#part-11-tagging-trials-with-additional-data) and [part 12](https://www.jspsych.org/7.0/tutorials/rt-task/#part-12-manipulating-data-during-the-experiment) of the simple RT task tutorial to see how to do these things. 

Your raven trials will end up looking something like this:
```js
var raven35 = {
  type: 'image-button-response',
  stimulus: 'Raven35_complete.png',
  choices: ['1', '2', '3', '4', '5', '6', '7', '8'],
  data: {
    task: 'raven',
    correct_response: '1'
  },
  on_finish: function(data) {
    // the value of "response" in the data will be the index of the button that was clicked, starting with 0
    // so we need to compare that number to the index of the correct response in the choices array ('1' = index 0, '2' = index 1, etc.)
    // if the two indices are the same, then "correct" will be true, otherwise it'll be false
    data.correct = data.response == ['1', '2', '3', '4', '5', '6', '7', '8'].indexOf(data.correct_response);
  }
};
```
(BTW I noticed some incorrect parameter names in the code you posted, so I've changed those above).

Your `show_results` trial looks good. The only thing there is that your set of raven trials is called `trials`, but then you switch to calling it `trial`. 

Also, if you want to check the values of your variables like `trials`, `correct_trials`, etc. you can put some `console.log`s inside the `stimulus` function, which will print the values to the browser's JavaScript console. This is useful for debugging problems. I've added some examples below.
```js
var show_results = {
  type: 'html-keyboard-response',
  stimulus: function () {
    var trials = jsPsych.data.get().filter({task: 'raven'});
    console.log('trials: ',trials); // this should print the data from all raven trials
    console.log('trial count: ', trials.count());  // this should print the total number of trials
    var correct_trials = trials.filter({correct: true});
    console.log('correct trial count: ', correct_trials.count());  // this should print the number of correct trials
    var accuracy = Math.round((correct_trials.count() / trials.count()) * 100);
    return `<p>You answered correctly in ${accuracy}% of the trials.</p>
         <p>Press any key to end the experiment. Thanks!</p> `;
  }
};
```
I hope this helps! Let us know if you get it working or run into any other problems.

**Reply by ler93** [reply]
Becky! It worked perfectly! Thank you so much for your help!! :)
